### PR TITLE
feat: metadata on dataset creation

### PR DESCRIPTION
**Pull request by mohammad-sdsc**
# Description

When creating datasets, users can provide its description, creators, and display name using command line options. Creator is a string with "Name <email>" format and users can pass multiple creators. 
This also makes a consistent use of `short_name` in the code. `short_name` is used as the dataset's data directory name and dataset's reference name. `name` is not used by Renku other than storing it as a metadata. If users don't provide a short name when creating a dataset then one is automatically created.

Fixes https://github.com/SwissDataScienceCenter/renku-python/issues/515
Fixes https://github.com/SwissDataScienceCenter/renku-python/issues/791
Fixes https://github.com/SwissDataScienceCenter/renku-python/issues/840

### CODE REVIEW

**Comment by emmjab** [comment]
I tried the following:
```
$ renku dataset create new -d "this is a test dataset" -c "Emma <e.jablonski@gmail.com>" --display-name "mountain"
OK
$ ls
Dockerfile		requirements.txt
```

(1) didn't realize that the `data` directory and a `new` or `mountain` subdirectory wasn't going to be created

```
(env) IC-LM-208782:my_new_project jablonsk$ renku dataset 
ID                                    DISPLAY_NAME    VERSION    CREATED              CREATORS
------------------------------------  --------------  ---------  -------------------  ----------
3041b30a-df24-4dcb-a0b9-709340433f88  mountain                   2019-12-04 14:31:11  Emma
(env) IC-LM-208782:my_new_project jablonsk$ renku dataset add new /Users/jablonsk/Documents/permafrosthackathon/permafrost-refresh/data/annotations/annotations.zip 
Error: Dataset "new" does not exist.
Use "renku dataset create new" to create the dataset or retry "renku dataset add new" command with "--create" option for automatic dataset creation.
```

(2) Didn't realize `display-name` was then going to be how I add files to my dataset & the name of the directory inside data.

```
(env) IC-LM-208782:my_new_project jablonsk$ renku dataset add mountain /Users/jablonsk/Documents/permafrosthackathon/permafrost-refresh/data/annotations/annotations.zip 
Warning: Adding data from local Git repository. Use remote's Git URL instead to enable lineage information and updates.
(env) IC-LM-208782:my_new_project jablonsk$ ls
Dockerfile		data			requirements.txt
```

- Also checked that the name & email get validated (i.e. has to be of the format "Name <email>")
- that you can pass in any combination of the options

I don't quite understand `display_name` -- what makes something a valid renku name?

**Comment by mohammad-sdsc** [comment]
1. Data directory is created when you add a file to the dataset.
2. `display_name` is used as a valid Renku name, however, its name is a bit misleading and we might change this.

**Comment by emmjab** [comment]
> 1. Data directory is created when you add a file to the dataset.

Not gonna dig into this one now; can see it going either way (i.e. if the directory shows up, users will think they can add files to the directory without `renku dataset add`-ing them, but if it doesn't show up it's kind of opaque about what happens when you "create a dataset".

> 2. `display_name` is used as a valid Renku name, however, its name is a bit misleading and we might change this.

What's the definition of a "valid Renku name"?

> 3. But then ls-files still shows the dataset being called new

What about the inconsistency between whether it's named by `renku dataset create <firstname>` v. `renku dataset create --display_name <lastname>`?

**Comment by mohammad-sdsc** [comment]
> Not gonna dig into this one now; can see it going either way (i.e. if the directory shows up, users will think they can add files to the directory without `renku dataset add`-ing them, but if it doesn't show up it's kind of opaque about what happens when you "create a dataset".

Please create an issue for this if you think we need to discuss it. 

> What's the definition of a "valid Renku name"?

A valid [Git reference](https://git-scm.com/docs/git-check-ref-format). Basically, one can use alphanumeric, ., -, and _. Some more characters are allowed by Git to use but we should (and will) disallow them.

> What about the inconsistency between whether it's named by `renku dataset create <firstname>` v. `renku dataset create --display_name <lastname>`?

The `--display-name` option is removed due the apparent confusion it causes. Now, there is only dataset's `name` (`<firstname>` in your example).

**Comment by rokroskar** [comment]
Since we are now using `internal_name`, could we also specify `--name` or `--title` to give it a more human-readable name?

File: renku/core/management/datasets.py [hunk 1]
```diff
@@ -174,6 +154,54 @@ def with_dataset(self, name=None, identifier=None, create=False):
 
         dataset.to_yaml()
 
+    def create_dataset(
+        self, name, internal_name='', description='', creators=()
```
Comment (rokroskar) [inline renku/core/management/datasets.py]: should allow for specifying an identifier?

**Comment by mohammad-sdsc** [comment]
> Since we are now using `internal_name`, could we also specify `--name` or `--title` to give it a more human-readable name?

We already have a `name` field for datasets which is the same as `internal_name` in locally-created datasets and is whatever-is-set-as-name for imported datasets. Probably, we should use a `title` field for human-readable names for both of these two types of datasets but in that case we are changing `name` field for imported datasets.

**Comment by rokroskar** [comment]
Right - I'm wondering if we could follow the same convention for newly-created datasets - if the name given by the user matches the "allowed characters" regex, then we keep it as is and `name` is the same as `internal_name`. However, if a user specifies the name in this way:

```
renku dataset create "Exploratory dataset"
```

then this would become

```
name -> "Exploratory dataset"
internal_name -> exploratory.dataset
```

**Comment by mohammad-sdsc** [comment]
I believe this will confuse some users as they expect to the name `"Exploratory dataset"` to refer to the dataset but they must actually use `exploratory.dataset`. This is the same situation as a previous commit of this PR that we had `--display-name` and a `name`.

**Comment by rokroskar** [comment]
What if we followed up the dataset creation with a helpful message that said something like:

```
Creation successful! Use the name exploratory.dataset to refer to this dataset.
```

We also need something like this to follow the import command anyway.

**Comment by mohammad-sdsc** [comment]
I had to force-push to resolve merge conflicts. Please review the last two commits.

**Comment by rokroskar** [comment]
Cool! I think that works nicely:

```
renku dataset create "My awesome dataset"
Use the name "my_awesome_dataset" to refer to this dataset.
OK
 rok   master  ❲c❳ base  ~/Projects/renku-test-projects/test-850  renku dataset
ID                                    INTERNAL_NAME             VERSION    CREATED              CREATORS
------------------------------------  ------------------------  ---------  -------------------  ---------------------------------------
8bdee124-1a8a-408e-9ff1-b98f9ba007c4  my_awesome_dataset                   2019-12-16 14:59:50  R.Roskar
```

@emmjab what do you think?

**Comment by emmjab** [comment]
Hah -- I think this is still confusing. Why does the name have to change? It has to be a valid git name? Why? it's not a submodule anymore, was that why?

**Comment by rokroskar** [comment]
Because it's annoying to type something that is super long and (potentially) contains special characters?

**Comment by rokroskar** [comment]
It's not obvious in this example, clearly, but if you import a dataset from zenodo it's likely that the name will be annoyingly long. You don't want to have to type that or copy/paste that every time you use it

**Comment by emmjab** [comment]
hmm... autocomplete? 😅 

> if you import a dataset from zenodo it's likely that the name will be annoyingly long

Why is that? I can google this

**Comment by rokroskar** [comment]
Take https://zenodo.org/record/3549866 for example. The name is "Synthetic dataset used in "The maximum weighted submatrix coverage problem: A CP approach"" - how do you suppose to use that on the command line?

**Comment by emmjab** [comment]
why not use the DOI?

**Comment by rokroskar** [comment]
Ok so @emmjab and I have been discussing this a bit offline. I find "internal_name" a bit confusing - what if we called it "short_name"? We could store this using `schema.org/alternateName` to give the user the option to set it or change it. The default would be what we are doing now, i.e. automatically generate it. 

So it would look like

```
dataset create --shortname dataset1 "This ia a dataset about stuff"
```

or 

```
dataset import --shortname dataset1 <DOI>
```

**Pull request merged**